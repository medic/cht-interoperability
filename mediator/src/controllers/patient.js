const axios = require('axios');
const { genereateFHIRPatientResource } = require('../utils/patient');
const FHIR_ROOT_URL = 'http://openhim-core:5001/fhir';

async function createPatient(CHTpatientDoc) {
  const FHITPatientResource = genereateFHIRPatientResource(CHTpatientDoc);

  try {
    const res = await axios.post(`${FHIR_ROOT_URL}/Patient`, FHITPatientResource, { auth: {
      username: 'interop-client',
      password: 'interop-password'
    }});
    return {status: res.status, patient: res.data}
  } catch (error) {
    console.error(error);
    return {status: error.status, patient: error.data}
  }
}

module.exports = {
  createPatient
};
