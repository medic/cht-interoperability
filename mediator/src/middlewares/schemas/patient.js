const joi = require('joi');

const createPatientSchema = joi.object({
  name: joi.string().optional(),
  _id: joi.string().optional(),
  sex: joi.string().optional(),
  birthDate: joi.string().optional(),
});

module.exports = {
  createPatientSchema,
};
