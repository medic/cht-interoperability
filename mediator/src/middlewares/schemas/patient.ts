import joi from "joi";


export const createPatientSchema = joi.object({
  parent: joi.any().optional(),
  type: joi.string().optional(),
  name: joi.string().required(),
  _id: joi.string().required(),
  id: joi.string().optional(),
  sex: joi.string().trim().valid('male', 'female', 'other', 'unkown').required(),
  // validate expecting YYYY-MM-DD
  date_of_birth: joi
    .string()
    .regex(/((?:19|20)\d\d)-(0?[1-9]|1[012])-([12][0-9]|3[01]|0?[1-9])/)
    .required()
    .messages({
      "object.regex": "Invalid date expecting YYYY-MM-DD",
      "string.pattern.base": "Invalid date expecting YYYY-MM-DD"
    })
});
