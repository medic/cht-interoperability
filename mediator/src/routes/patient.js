const {Router} = require('express');
const router = Router();
const patientController = require('../controllers/patient');
const {validateBodyAgainst} = require('../middlewares');
const {createPatientSchema} = require('../middlewares/schemas/patient');

router.post('/',
  validateBodyAgainst(createPatientSchema),
  async function(req, res) {
    const {status, patient} = await patientController.createPatient(req.body);
    res.send({status, patient});
});

module.exports = router;
