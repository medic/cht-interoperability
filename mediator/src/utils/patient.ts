function genereateFHIRPatientResource(patient: any) {
  const patientLastName = patient.name.split(' ').slice(-1);
  const FHITPatientResource = {
    resourceType: 'Patient',
    id: patient.id,
    identifier: [
      {
        system: 'cht',
        value: patient._id
      }
    ],
    name: [
      {
        use: 'official',
        family: patientLastName,
        given: [patient.name]
      }
    ],
    gender: patient.sex,
    birthDate: patient.date_of_birth
  };

  return FHITPatientResource;
}

module.exports = {
  genereateFHIRPatientResource,
};
