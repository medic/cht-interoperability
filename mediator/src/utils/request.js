function requestHandler(handler) {
  return  (req, res) => {
    handler(req)
    .then(({status, data}) => {
      res.status(status).send(data);
    });
  };
}

module.exports = {
  requestHandler,
};
