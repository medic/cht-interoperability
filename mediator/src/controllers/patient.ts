import axios from 'axios';
import {generateFHIRPatientResource} from '../utils/patient';
import {FHIR} from '../../config';
import {logger} from '../../logger';
const {url: fhirUrl, username: fhirUsername, password: fhirPassword} = FHIR;

export async function createPatient(CHTpatientDoc: any) {
  try {
    const FHIRPatientResource = generateFHIRPatientResource(CHTpatientDoc);
    const res = await axios.post(`${fhirUrl}/Patient`, FHIRPatientResource, {auth: {
      username: fhirUsername,
      password: fhirPassword,
    }});
    return {status: res.status, patient: res.data};
  } catch (error: any) {
    logger.error(error);

    if (!error.status) {
      return {status: 400, patient: {message: error.message}};
    }

    return {status: error.status, patient: error.data};
  }
}
