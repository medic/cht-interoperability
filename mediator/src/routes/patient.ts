import { Router } from "express";
import { createPatient } from "../controllers/patient";
import { validateBodyAgainst } from "../middlewares";
import { createPatientSchema } from "../middlewares/schemas/patient";
import { requestHandler } from "../utils/request";

const router = Router();

router.post(
  "/",
  validateBodyAgainst(createPatientSchema),
  requestHandler((req) => createPatient(req.body))
);

export default router;
