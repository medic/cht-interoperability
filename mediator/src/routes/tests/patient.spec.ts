import request from "supertest";
import app from "../../..";
import { createPatient } from "../../controllers/patient";

jest.mock("../../controllers/patient");

describe("POST /patient", () => {
  it("calls handler for valid incoming request", async () => {
    (createPatient as any).mockResolvedValueOnce({
      data: {},
      status: 201,
    });

    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      id: "OPTIONAL",
      sex: "male",
      date_of_birth: "2000-01-01",
      parent: "OPTIONAL",
      type: "OPTIONAL",
    };

    const res = await request(app).post("/patient").send(data);

    expect(res.status).toBe(201);
    expect(res.body).toEqual({});
    expect(createPatient).toHaveBeenCalledWith(data);
    expect(createPatient).toHaveBeenCalled();
    expect(createPatient).toMatchSnapshot();
  });

  it("returns a bad request error for invalid incoming request", async () => {
    (createPatient as any).mockResolvedValueOnce({
      data: {},
      status: 201,
    });

    const data = {
      name: "John Doe",
      _id: "JOHN_DOE_ID",
      id: "OPTIONAL",
      sex: "male",
      date_of_birth: "WRONG_DATE",
      parent: "OPTIONAL",
      type: "OPTIONAL",
    };

    const res = await request(app).post("/patient").send(data);

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
    expect(createPatient).not.toHaveBeenCalled();
  });
});
