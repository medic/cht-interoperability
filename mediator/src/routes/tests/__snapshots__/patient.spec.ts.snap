// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /patient calls handler for valid incoming request 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "_id": "JOHN_DOE_ID",
        "date_of_birth": "2000-01-01",
        "id": "OPTIONAL",
        "name": "John Doe",
        "parent": "OPTIONAL",
        "sex": "male",
        "type": "OPTIONAL",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`POST /patient returns a bad request error for invalid incoming request 1`] = `
{
  "message": "Invalid date expecting YYYY-MM-DD",
}
`;
